packages/effect/src/internal/fiberRuntime.ts

```diff
@@ -2631,111 +2630,61 @@ export const raceAll: <Eff extends Effect.Effect<any, any, any>>(
-                        ))
-                    )
-                  ),
-                  core.flatMap((fibers) =>
-                    pipe(
-                      restore(pipe(Deferred.await(done), core.flatMap(inheritAll))),
-                      core.onInterrupt(() =>
-                        pipe(
-                          fibers,
-                          RA.reduce(
-                            core.void,
-                            (effect, fiber) => pipe(effect, core.zipLeft(core.interruptFiber(fiber)))
-                          )
-                        )
-                      )
-                    )
-                  )
-                )
-              )
-            )
-          )
+  return core.uninterruptibleMask((restore) =>
+    core.withFiberRuntime((state, status) => {
+      const fibers = Array.from(all).map((self) =>
+        unsafeFork(
+          core.interruptible(self),
+          state,
+          status.runtimeFlags
         )
       )
-    )
-  )
-}
-
-const raceAllArbiter = <E, E1, A, A1>(
-  fibers: Iterable<Fiber.Fiber<A | A1, E | E1>>,
-  winner: Fiber.Fiber<A | A1, E | E1>,
-  deferred: Deferred.Deferred<readonly [A | A1, Fiber.Fiber<A | A1, E | E1>], E | E1>,
-  fails: Ref.Ref<number>
-) =>
-(exit: Exit.Exit<A | A1, E | E1>): Effect.Effect<void> =>
-  core.exitMatchEffect(exit, {
-    onFailure: (cause) =>
-      pipe(
-        Ref.modify(fails, (fails) =>
-          [
-            fails === 0 ?
-              pipe(core.deferredFailCause(deferred, cause), core.asVoid) :
-              core.void,
-            fails - 1
-          ] as const),
-        core.flatten
-      ),
-    onSuccess: (value): Effect.Effect<void> =>
-      pipe(
-        core.deferredSucceed(deferred, [value, winner] as const),
-        core.flatMap((set) =>
-          set ?
-            pipe(
-              Chunk.fromIterable(fibers),
-              RA.reduce(
-                core.void,
-                (effect, fiber) =>
-                  fiber === winner ?
-                    effect :
-                    pipe(effect, core.zipLeft(core.interruptFiber(fiber)))
-              )
-            ) :
-            core.void
+      if (fibers.length === 0) {
+        return core.dieSync(() => new core.IllegalArgumentException(`Received an empty collection of effects`))
+      }
+      let winner: Fiber.Fiber<A, E>
+      const interruptAll = () => {
+        for (const fiber of fibers) {
+          if (fiber !== winner) {
+            fiber.unsafeInterruptAsFork(state.id())
+          }
+        }
+      }
+      return restore(core.async<A, E, R>((resume) => {
+        let failed = 0
+        for (const fiber of fibers) {
+          fiber.addObserver((exit) => {
+            if (exit._tag === "Success") {
+              winner = fiber
+              interruptAll()
+              resume(exit)
+            } else {
+              failed++
+              if (failed === fibers.length) {
+                resume(exit)
```

Comment:
Should this be a Cause.Parallel?

Reply:
You can also remove the mask and do everything in a single observer: https://github.com/Effect-TS/effect/pull/5449